samples/subsys/usb/hid-cdc/src/main.c

```diff
@@ -0,0 +1,868 @@
+	} else if (ascii < 128) {
+		/* Special characters #4 */
+		switch (ascii) {
+		case 123:
+			return HID_KEY_LEFTBRACE;
+		case 124:
+			return HID_KEY_BACKSLASH;
+		case 125:
+			return HID_KEY_RIGHTBRACE;
+		case 126:
+			return HID_KEY_GRAVE;
+		case 127:
+			return HID_KEY_DELETE;
+		default:
+			return -1;
+		}
+	}
+
+	return -1;
+}
+
+static inline bool needs_shift(u8_t ascii)
+{
+	if ((ascii < 33) || (ascii == 39)) {
+		return false;
+	} else if ((ascii >= 33) && (ascii < 44)) {
+		return true;
+	} else if ((ascii >= 44) && (ascii < 58)) {
+		return false;
+	} else if ((ascii == 59) || (ascii == 61)) {
+		return false;
+	} else if ((ascii >= 58) && (ascii < 91)) {
+		return true;
+	} else if ((ascii >= 91) && (ascii < 94)) {
+		return false;
+	} else if ((ascii == 94) || (ascii == 95)) {
+		return true;
+	} else if ((ascii > 95) && (ascii < 123)) {
+		return false;
+	} else if ((ascii > 122) && (ascii < 127)) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+/* CDC ACM */
+
+static volatile bool data_transmitted;
+static volatile bool data_arrived;
+
+static void flush_buffer_mouse(void)
+{
+	chr_ptr_mouse = 0;
+	memset(data_buf_mouse, 0, sizeof(data_buf_mouse));
+}
+
+static void flush_buffer_kbd(void)
+{
+	chr_ptr_kbd = 0;
+	memset(data_buf_kbd, 0, sizeof(data_buf_kbd));
+}
+
+static void write_data(struct device *dev, const char *buf, int len)
+{
+	uart_irq_tx_enable(dev);
+
+	while (len) {
+		int written;
+
+		data_transmitted = false;
+		written = uart_fifo_fill(dev, (const u8_t *)buf, len);
+		while (data_transmitted == false) {
+			k_yield();
+		}
+
+		len -= written;
+		buf += written;
+	}
+
+	uart_irq_tx_disable(dev);
+}
+
+static void cdc_mouse_int_handler(struct device *dev)
+{
+	uart_irq_update(dev);
+
+	if (uart_irq_tx_ready(dev)) {
+		data_transmitted = true;
+	}
+
+	if (uart_irq_rx_ready(dev)) {
+		u32_t bytes_read;
+
+		while ((bytes_read = uart_fifo_read(dev,
+		       (u8_t *)data_buf_mouse+chr_ptr_mouse,
+		       sizeof(data_buf_mouse)-chr_ptr_mouse))) {
+			chr_ptr_mouse += bytes_read;
+			if (data_buf_mouse[chr_ptr_mouse-1] == 0x0D) {
+				/* ENTER */
+				struct app_evt_t *ev = app_evt_alloc();
+
+				data_buf_mouse[chr_ptr_mouse-1] = '\0';
+
+				if (!strcmp(data_buf_mouse, "up")) {
+					ev->event_type = CDC_UP;
+				} else if (!strcmp(data_buf_mouse, "down")) {
+					ev->event_type = CDC_DOWN;
+				} else if (!strcmp(data_buf_mouse, "right")) {
+					ev->event_type = CDC_RIGHT;
+				} else if (!strcmp(data_buf_mouse, "left")) {
+					ev->event_type = CDC_LEFT;
+				} else {
+					ev->event_type = CDC_UNKNOWN;
+				}
+				flush_buffer_mouse();
+				app_evt_put(ev);
+				k_sem_give(&evt_sem);
+			}
+
+			if (chr_ptr_mouse >= sizeof(data_buf_mouse)) {
+				LOG_WRN("Buffer overflow");
+				flush_buffer_mouse();
+			}
+		}
+	}
+}
+
+static void cdc_kbd_int_handler(struct device *dev)
+{
+	uart_irq_update(dev);
+
+	if (uart_irq_tx_ready(dev)) {
+		data_transmitted = true;
+	}
+
+	if (uart_irq_rx_ready(dev)) {
```

Comment:
Perhaps, it'd be more readable if you `return` early from this function, when the `rx` interrupt is not ready.

Reply:
If you mean:
```
if(!uart_irq_rx_ready(dev)) {
return;
} else {
...
}
```
I don't think it will improve readability.

Reply:
Fixed.